net_unix.go

```diff
@@ -0,0 +1,225 @@
+	netUnixStateConnecting   = 2
+	netUnixStateConnected    = 3
+	netUnixStateDisconnected = 4
+)
+
+// NetUnixType is the type of the type field
+type NetUnixType uint64
+
+// NetUnixFlags is the type of the flags field
+type NetUnixFlags uint64
+
+// NetUnixState is the type of the state field
+type NetUnixState uint64
+
+// NetUnixLine represents a line of /proc/net/unix
+type NetUnixLine struct {
+	KernelPtr string
+	RefCount  uint64
+	Protocol  uint64
+	Flags     NetUnixFlags
+	Type      NetUnixType
+	State     NetUnixState
+	Inode     string
+	Path      string
+}
+
+// NetUnix holds the data read from /proc/net/unix
+type NetUnix struct {
+	Rows []*NetUnixLine
+}
+
+// NewNetUnix returns data read from /proc/net/unix.
+func NewNetUnix() (*NetUnix, error) {
+	fs, err := NewFS(DefaultMountPoint)
+	if err != nil {
+		return nil, err
+	}
+
+	return fs.NewNetUnix()
+}
+
+// NewNetUnix returns data read from /proc/net/unix.
+func (fs FS) NewNetUnix() (*NetUnix, error) {
+	return NewNetUnixByPath(fs.Path("net/unix"))
+}
+
+// NewNetUnixByPath returns data read from /proc/net/unix by file path.
+// It might returns an error with partiail parsed data, if an error occur after some data parsed
+func NewNetUnixByPath(path string) (*NetUnix, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	nu := &NetUnix{
+		Rows: make([]*NetUnixLine, 0, 32),
+	}
+	scanner := bufio.NewScanner(f)
+	// omit the header line
+	scanner.Scan()
+	header := scanner.Text()
+	// it only contains Inode field in kernel version older than 3.2
+	hasInode := strings.Contains(header, "Inode")
+	for scanner.Scan() {
+		line := scanner.Text()
+		item, err := nu.parseLine(line, hasInode)
+		if err != nil {
+			return nu, err
+		}
+		nu.Rows = append(nu.Rows, item)
+	}
+
+	return nu, nil
+}
+
+func (u *NetUnix) parseLine(line string, hasInode bool) (*NetUnixLine, error) {
+	fields := strings.Fields(line)
+	users, err := u.parseUsers(fields[netUnixRefCountIdx])
+	if err != nil {
+		return nil, fmt.Errorf("Parse Unix domain ref count(%s) failed: %s", fields[netUnixRefCountIdx], err)
+	}
+	flags, err := u.parseFlags(fields[netUnixFlagsIdx])
+	if err != nil {
+		return nil, fmt.Errorf("Parse Unix domain flags(%s) failed: %s", fields[netUnixFlagsIdx], err)
+	}
+	typ, err := u.parseType(fields[netUnixTypeIdx])
+	if err != nil {
+		return nil, fmt.Errorf("Parse Unix domain type(%s) failed: %s", fields[netUnixTypeIdx], err)
+	}
+	state, err := u.parseState(fields[netUnixStateIdx])
+	if err != nil {
+		return nil, fmt.Errorf("Parse Unix domain state(%s) failed: %s", fields[netUnixStateIdx], err)
+	}
+
+	nuLine := &NetUnixLine{
+		KernelPtr: u.parseKernelPtr(fields[netUnixKernelPtrIdx]),
+		RefCount:  users,
+		Type:      typ,
+		Flags:     flags,
+		State:     state,
+		Path:      fields[len(fields)-1],
+	}
+	if hasInode {
+		nuLine.Inode = fields[len(fields)-2]
+	}
+	return nuLine, nil
+}
+
+func (u NetUnix) parseKernelPtr(str string) string {
```

Comment:
Shouldn't this function also return an error if this field doesn't end with a `:`? According to `af_unix.c#L2815`, it should always.